### PR TITLE
Push services 4220

### PR DESCRIPTION
**Pull request by dmitrytrager**
friend_checkin

According to: https://redmine.smartdata.solutions/issues/4220

### CODE REVIEW

File: app/controllers/users/checkins_controller.rb [hunk 1]
```diff
@@ -1,29 +1,28 @@
 class Users::CheckinsController < ApplicationController
   protect_from_forgery except: :show
+
   before_action :authenticate_user!
   before_action :require_checkin_ownership, except: [:index, :new, :create, :destroy_all]
   before_action :require_device_ownership, only: [:index, :new, :create, :destroy_all]
 
   def new
-    @device = Device.find(params[:device_id])
```
Comment (dmitrytrager) [inline app/controllers/users/checkins_controller.rb]: it was obviously redundant here to do same thing twice

File: app/controllers/users/checkins_controller.rb [hunk 2]
```diff
@@ -57,6 +56,10 @@ def destroy_all
 
   private
 
+  def device
```
Comment (dmitrytrager) [inline app/controllers/users/checkins_controller.rb]: small optimization here

File: app/controllers/users/checkins_controller.rb [hunk 3]
```diff
@@ -1,29 +1,28 @@
...
+      total: device.checkins.count
     }
   end
 
   def create
-    @device = Device.find(params[:device_id])
-    @checkin = @device.checkins.create(allowed_params)
-    @device.notify_subscribers('new_checkin', @checkin)
+    @checkin = device.checkins.create(allowed_params)
+    NotifyAboutCheckin.call(device: device, checkin: @checkin)
```
Comment (dmitrytrager) [inline app/controllers/users/checkins_controller.rb]: introduced interactor here

File: spec/helpers/friends_helper_spec.rb [hunk 4]
```diff
@@ -1,12 +1,10 @@
...
-  let(:friend) { FactoryGirl.create :user }
-  let(:device) do
-    device = FactoryGirl.create(:device, user_id: friend.id)
-    device.checkins << [FactoryGirl.create(:checkin), FactoryGirl.create(:checkin)]
-    device
-  end
+  let(:friend) { create :user }
+  let(:device) { create :device, user: friend }
+
+  before { create_list :checkin, 2, device: device, city: "city" }
```
Comment (dmitrytrager) [inline spec/helpers/friends_helper_spec.rb]: This is a better way to work with associations.
Other cases can be updated accordingly. It's easier to describe explicitly when we do not need association (like [here](https://github.com/earlymarket/Coposition/pull/634/files#diff-1a0461c5944f6bf97b22ff9bb9b63ec7R6)).

File: spec/helpers/friends_helper_spec.rb [hunk 5]
```diff
@@ -15,7 +13,7 @@
 
     it 'returns the last checkin address if it exists' do
       expect(helper.friends_device_last_checkin(device.checkins.first)).to match('Last available')
-      expect(helper.friends_device_last_checkin(device.checkins.first)).to match(device.checkins.first.fogged_city.to_s)
+      expect(helper.friends_device_last_checkin(device.checkins.first)).to match(device.checkins.first.city.to_s)
```
Comment (dmitrytrager) [inline spec/helpers/friends_helper_spec.rb]: Looks like I found a small bug here. B/c helper invokes methods that work with 'city' attribute, not fogged. Please, confirm.

Comment (t-harps) [inline spec/helpers/friends_helper_spec.rb]: Depends on the checkin. The checkin city will sometimes have been set to 'fogged_city' if the device is fogged and the friend can not bypass fogging. However usually the fogged_city == city anyway. I think in this test though the device is fogged so hence why we use fogged_city.

**Comment by t-harps** [comment]
@dmitrytrager is this finished? Struggling to get this to work locally/on staging, could you advise.

**Comment by t-harps** [comment]
@dmitrytrager Is additional config required to set up Redis on staging/production?

File: spec/helpers/friends_helper_spec.rb (hunk 5 above)
Comment (dmitrytrager) [inline spec/helpers/friends_helper_spec.rb]: No action except helper here. And it [operates](https://github.com/earlymarket/Coposition/blob/363f41dbd8e01585f7fb8a414c697d8aa29979de/app/helpers/friends_helper.rb#L4) with 'city' attribute @t-harps .

Comment (t-harps) [inline spec/helpers/friends_helper_spec.rb]: The checkin that is passed to the helper is not a regular checkin object though. As this is the friends page, the checkin will have been returned from safe_checkin_info_for, in fact this method should probably be rewritten as follows:
```
def friends_device_last_checkin(device)
  checkin = device.safe_checkin_info_for(permissible: current_user, action: 'last')[0]
  if checkin.present?
    "<p>Last available reported area: #{checkin['city']}</p>".html_safe
  else
    '<p>No location found</p>'.html_safe
  end
end
```

In safe_checkin_info_for it is possible for the checkin.city to be replaced with fogged_city.

Comment (dmitrytrager) [inline spec/helpers/friends_helper_spec.rb]: Yep, but in test it is a simple checkin object, right?

Comment (t-harps) [inline spec/helpers/friends_helper_spec.rb]: Yup, so probably not a very good test at the moment as method should never actually receive a simple checkin object.

Comment (dmitrytrager) [inline spec/helpers/friends_helper_spec.rb]: @t-harps I'll leave it for now just in order not to drop test coverage, ok?

Comment (t-harps) [inline spec/helpers/friends_helper_spec.rb]: Yup that's fine, we have some tickets which are more focused on improving specs anyway.

**Comment by dmitrytrager** [comment]
@t-harps I've just added addon for redis. Now review app is deployed and we can check.
Need to keep in mind that feature should work only on the pages when friend's checkins are visible (like on a screenshot).

**Comment by t-harps** [comment]
@dmitrytrager Still having trouble with this, for some reason trying to create a new device gives me an error screen, so I created one using the heroku rails console and added a friend. When I try and create a checkin on this device the page does not update for either the user creating the checkin or the friend viewing a page on which the friends checkins are visible.

The heroku logs do not seem to give any useful information.

**Comment by dmitrytrager** [comment]
@t-harps will check this out. Can you provide error screen?

**Comment by dmitrytrager** [comment]
@t-harps updated review app (there was a typo in app.json)
But still want to see error screen.

**Comment by t-harps** [comment]
@dmitrytrager Error screen no longer appearing, that change must have fixed the issue.

**Comment by dmitrytrager** [comment]
Cool!

File: config/environments/staging.rb [hunk 6]
```diff
@@ -41,6 +41,15 @@
   # config.action_dispatch.x_sendfile_header = 'X-Sendfile' # for Apache
   # config.action_dispatch.x_sendfile_header = 'X-Accel-Redirect' # for NGINX
 
+  # Action Cable endpoint configuration
+  config.action_cable.url = "wss://coposition-staging-pr-634.herokuapp.com/cable"
```
Comment (dmitrytrager) [inline config/environments/staging.rb]: ATTENTION:
will need to update before merging

File: config/initializers/assets.rb [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # Be sure to restart your server when you modify this file.
 
 # Version of your assets, change this if you want to expire all your assets.
-Rails.application.config.assets.version = '1.00'
+Rails.application.config.assets.version = '1.01'
```
Comment (dmitrytrager) [inline config/initializers/assets.rb]: Actioncable refused to work on staging. Actually, to be included in application.js. I had to update assets version to fix this.

**Comment by dmitrytrager** [comment]
@t-harps need you help with two things:
1. I have user with friend, friend's device and checkins, but see nothing on friends page. Do I need to set specific attribute for some of these model?
2. Please try review app when you have some time and tell me, if you are able to see it's working.

**Comment by t-harps** [comment]
@dmitrytrager 
1. How did you create the friendship? I can see there are two approvals between the admin user and another user, but these have no status or approval date. It also looks like no permission was created (device -> friend relationship) which determines what info from a device the friend can see.

2. Trying now.

**Comment by dmitrytrager** [comment]
Cool, thanks

**Comment by dmitrytrager** [comment]
@t-harps please, let me know if you have no more questions for this PR

**Comment by t-harps** [comment]
@dmitrytrager Looks good. I think what I'll do is keep this PR open and try and do a bit of additional work on this branch myself to familiarise my self with the code. I will try adding the Notify to API checkin create and also make sure permissions are taken into account before 'notifying' a friend of a new checkin.

**Comment by dmitrytrager** [comment]
Okay

File: config/environments/staging.rb (hunk 6 above)
Comment (t-harps) [inline config/environments/staging.rb]: @dmitrytrager Thinking about merging this soon, what needs to be updated?

File: app/assets/javascripts/channels/friend.es6 [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+  // disconnected: function() {
+  // },
+
+  received: function(data) {
+    switch (data.action) {
+      case "checkin":
+        if ($(".c-friends.a-show_device").length === 0) { return };
+
+        if(data.privilege === 'complete'){
```
Comment (dmitrytrager) [inline app/assets/javascripts/channels/friend.es6]: missing space after `if`

File: app/assets/javascripts/channels/friend.es6 [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        } else {
+          gon.checkins = [data.msg];
+        }
+
+        COPO.maps.refreshMarkers(gon.checkins);
+
+        break;
+
+      case "destroy":
+        if ($(".c-friends.a-show_device").length === 0) { return };
```
Comment (dmitrytrager) [inline app/assets/javascripts/channels/friend.es6]: Do you avoid type conversion intentionally when using '===' ?

File: app/controllers/users/checkins_controller.rb [hunk 10]
```diff
@@ -46,6 +45,8 @@ def update
 
   def destroy
     @checkin = Checkin.find_by(id: params[:id]).delete
+    NotifyAboutDestroyCheckin.call(device: device, checkin: @checkin)
```
Comment (dmitrytrager) [inline app/controllers/users/checkins_controller.rb]: :+1:

File: config/environments/staging.rb (hunk 6 above)
Comment (dmitrytrager) [inline config/environments/staging.rb]: @t-harps this url should point to staging, not to review app

File: app/assets/javascripts/channels/friend.es6 (hunk 9 above)
Comment (t-harps) [inline app/assets/javascripts/channels/friend.es6]: Yes, jslint recommends it as apparently using '==' you can get some weird cases where things don't equal what you'd expect. Just using === everywhere to be consistent.

**Comment by dmitrytrager** [comment]
looks nice

**Comment by GarySiu** [comment]
Looks good. Don't know why travis was having problems previously.

**Pull request merged**